api/xds_protocol.rst

```diff
@@ -36,43 +36,122 @@ occurs.
+Every configuration resource in the xDS API has a type associated with it. The following types are supported:
 
-Type URLs
-^^^^^^^^^
-
-Each xDS API is concerned with resources of a given type. There is a 1:1
-correspondence between an xDS API and a resource type. That is:
-
--  LDS: :ref:`envoy.api.v2.Listener <envoy_api_msg_Listener>`
--  RDS: :ref:`envoy.api.v2.RouteConfiguration <envoy_api_msg_RouteConfiguration>`
--  VHDS: :ref:`envoy.api.v2.Vhds <envoy_api_msg_RouteConfiguration>`
--  CDS: :ref:`envoy.api.v2.Cluster <envoy_api_msg_Cluster>`
--  EDS: :ref:`envoy.api.v2.ClusterLoadAssignment <envoy_api_msg_ClusterLoadAssignment>`
--  SDS: :ref:`envoy.api.v2.Auth.Secret <envoy_api_msg_Auth.Secret>`
--  RTDS: :ref:`envoy.service.discovery.v2.Runtime <envoy_api_msg_service.discovery.v2.Runtime>`
+-  :ref:`envoy.api.v2.Listener <envoy_api_msg_Listener>`
+-  :ref:`envoy.api.v2.RouteConfiguration <envoy_api_msg_RouteConfiguration>`
+-  :ref:`envoy.api.v2.route.VirtualHost <envoy_api_msg_route.VirtualHost>`
+-  :ref:`envoy.api.v2.Cluster <envoy_api_msg_Cluster>`
+-  :ref:`envoy.api.v2.ClusterLoadAssignment <envoy_api_msg_ClusterLoadAssignment>`
+-  :ref:`envoy.api.v2.Auth.Secret <envoy_api_msg_Auth.Secret>`
+-  :ref:`envoy.service.discovery.v2.Runtime <envoy_api_msg_service.discovery.v2.Runtime>`
 
 The concept of `type URLs <https://developers.google.com/protocol-buffers/docs/proto3#any>`_ appears below, and takes the form
-`type.googleapis.com/<resource type>`, e.g.
-`type.googleapis.com/envoy.api.v2.Cluster` for CDS. In various
-requests from Envoy and responses by the management server, the resource
-type URL is stated.
+`type.googleapis.com/<resource type>` -- e.g., `type.googleapis.com/envoy.api.v2.Cluster` for a `Cluster` resource. In various
+requests from Envoy and responses by the management server, the resource type URL is stated.
+
+Variants of the xDS API
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Four Variants
+^^^^^^^^^^^^^
+
+There are four variants of the xDS API used via streaming gRPC, which cover all combinations of two dimensions.
+
+The first dimension is State of the World (SotW) vs. incremental. The SotW approach was the original mechanism used by xDS, in
+which the client must specify all resource names it is interested in with each request (except when making a wildcard
+request in LDS/CDS), and the server must return all resources the client has subscribed to in each request (in LDS/CDS).
+This means that if the client is already subscribing to 99 resources and wants to add an additional one, it must send a request
+with all 100 resource names, rather than just the one new one. And the server must then respond by sending all 100 resources,
+even if the 99 that were already subscribed to have not changed (in LDS/CDS). This mechanism can be a scalability limitation,
+which is why the incremental API variant was introduced. The incremental approach allows both the client and server to indicate
```

Comment:
For the incremental approach, shouldn't it only applies to responses? That is, the client should always send requests containing the full set of resources it is looking for while the server replies with multiple responses with each containing only a subset of requested resources. 

For example, if client initially requests resource A. It sends a request for A and server responds A. Later the client want B as well, shouldn't it send a request for A and B? The server responds only B.

Reply:
No.  In the incremental API, the request type is `DeltaDiscoveryRequest` instead of `DiscoveryRequest`, so instead of a single `resource_names` field, there are two fields: `resource_names_subscribe` and `resource_names_unsubscribe`.  So the client can express adds or deletes to the set of resource names it is interested in.

See the "How the client specifies what resources to return" section below for details.

Reply:
Oh, you were talking about `DeltaDiscoveryRequest`. What I am mentioning here is for RDS/EDS, which are inherently incremental.

Reply:
RDS and EDS are not really incremental, at least not in the sense in which the term "incremental" is used in this document.  They are similar to the incremental API in the sense that each resource can be sent in its own response (as described in the "Grouping Resources into Responses" section below), but their request semantics are still SotW-style.